packages/swapper/src/types.ts

```diff
@@ -36,3 +61,294 @@ export enum SwapErrorType {
+
+export type CommonGetUnsignedTransactionArgs = {
+  tradeQuote: TradeQuote
+  chainId: ChainId
+  stepIndex: number
+  slippageTolerancePercentageDecimal: string
+  supportsEIP1559?: boolean
+}
+
+export type GetUnsignedEvmTransactionArgs = CommonGetUnsignedTransactionArgs & EvmAccountMetadata
+export type GetUnsignedEvmMessageArgs = GetUnsignedEvmTransactionArgs
+export type GetUnsignedUtxoTransactionArgs = CommonGetUnsignedTransactionArgs & UtxoAccountMetadata
+export type GetUnsignedCosmosSdkTransactionArgs = CommonGetUnsignedTransactionArgs &
+  CosmosSdkAccountMetadata
+
+// the client should never need to know anything about this payload, and since it varies from
+// swapper to swapper, the type is declared this way to prevent generics hell while ensuring the
+// data originates from the correct place (assuming no casting).
+export type UnsignedTx = Nominal<Record<string, any>, 'UnsignedTx'>
+
+export type ExecuteTradeArgs = {
+  senderAddress: string
+  receiverAddress: string
+  txToSign: UnsignedTx
+  wallet: HDWallet
+  chainId: ChainId
+}
+
+export type ExecuteTradeArgs2 = {
+  txToSign: UnsignedTx
+  wallet: HDWallet
+  chainId: ChainId
+}
+
+export type CheckTradeStatusInput = {
+  quoteId: string
+  txHash: string
+  chainId: ChainId
+  stepIndex: number
+}
+
+// a result containing all routes that were successfully generated, or an error in the case where
+// no routes could be generated
+type TradeQuoteResult = Result<TradeQuote[], SwapErrorRight>
+
+export type EvmTransactionRequest = {
+  gasLimit: string
+  to: string
+  from: string
+  value: string
+  data: string
+  chainId: number
+} & evm.types.Fees
+
+export type CowMessageToSign = {
+  chainId: ChainId
+  orderToSign: CowSwapOrder
+}
+
+// TODO: one day this might be a union to support various implementations or generic 💀
+export type EvmMessageToSign = CowMessageToSign
+
+export type Swapper = {
+  filterAssetIdsBySellable: (assets: Asset[]) => Promise<AssetId[]>
+  filterBuyAssetsBySellAssetId: (input: BuyAssetBySellIdInput) => Promise<AssetId[]>
+  executeTrade?: (executeTradeArgs: ExecuteTradeArgs) => Promise<string>
+
+  executeEvmTransaction?: (
+    txToSign: EvmTransactionRequest,
+    callbacks: EvmTransactionExecutionProps,
+  ) => Promise<string>
+  executeEvmMessage?: (
+    txMetaToSign: EvmMessageToSign,
+    callbacks: EvmMessageExecutionProps,
+  ) => Promise<string>
+  executeUtxoTransaction?: (
+    txToSign: BTCSignTx,
+    callbacks: UtxoTransactionExecutionProps,
+  ) => Promise<string>
+  executeCosmosSdkTransaction?: (
+    txToSign: StdSignDoc,
+    callbacks: CosmosSdkTransactionExecutionProps,
+  ) => Promise<string>
+}
+
+export type SwapperApi = {
+  checkTradeStatus: (
+    input: CheckTradeStatusInput,
+  ) => Promise<{ status: TxStatus; buyTxHash: string | undefined; message: string | undefined }>
+  getTradeQuote: (input: GetTradeQuoteInput, assetsById: AssetsById) => Promise<TradeQuoteResult>
```

Comment:
This was partial before
```suggestion
  getTradeQuote: (input: GetTradeQuoteInput, assetsById: AssetsByIdPartial) => Promise<TradeQuoteResult>
```